### PR TITLE
feat(ingestion): extend feast plugin to ingest tags and owners

### PR DESCRIPTION
**Pull request by margaridafernandes-trip**
## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [x] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/scripts/datahub_preflight.sh [hunk 1]
```diff
@@ -45,7 +45,7 @@ arm64_darwin_preflight() {
   	pip3 install --no-use-pep517 scipy
   fi
 
-  brew_install "openssl@1.1"
+  brew_install "openssl@3.0.14"
```
Comment (jjoyce0510) [inline metadata-ingestion/scripts/datahub_preflight.sh:48]: Quick question - Why this change?

File: metadata-ingestion/scripts/datahub_preflight.sh [hunk 2]
```diff
@@ -1,4 +1,4 @@
-#!/bin/bash -e
+ #!/bin/bash -e
```
Comment (jjoyce0510) [inline metadata-ingestion/scripts/datahub_preflight.sh:1]: Why this space change?

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 3]
```diff
@@ -296,12 +330,29 @@ def _get_feature_view_workunit(self, feature_view: FeatureView) -> MetadataWorkU
...
 
+        aspects = [
+            BrowsePathsClass(paths=[f"/feast/{self.feature_store.project}"]),
+            StatusClass(removed=False),
+        ]
+
+        if feature_view.tags.get("name"):
+            tag = feature_view.tags.get("name")
+            tag_association = TagAssociationClass(tag=builder.make_tag_urn(tag))
+            global_tags_aspect = GlobalTagsClass(tags=[tag_association])
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:341]: Just note: If you've attached tags in DataHub, this will replace them by default :)

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 4]
```diff
@@ -296,12 +330,29 @@ def _get_feature_view_workunit(self, feature_view: FeatureView) -> MetadataWorkU
...
+            aspects.append(global_tags_aspect)
+
+        if feature_view.owner:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:344]: Just note: If you've attached owners in DataHub UI, this will replace them by default :)

File: metadata-ingestion/scripts/datahub_preflight.sh (hunk 1 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/scripts/datahub_preflight.sh:48]: I needed to upgrade the dependency version because the 1.1 version was deprecated.

File: metadata-ingestion/scripts/datahub_preflight.sh (hunk 2 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/scripts/datahub_preflight.sh]: fixed

File: metadata-ingestion/scripts/datahub_preflight.sh (hunk 1 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/scripts/datahub_preflight.sh]: I reverted to the previous version

File: metadata-ingestion/src/datahub/ingestion/source/feast.py (hunk 4 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py]: See comment above

File: metadata-ingestion/src/datahub/ingestion/source/feast.py (hunk 3 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py]: The goal is to have Feast as our source of truth, so the fields must be defined in Feast repo. 
Do you agree with this?

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 5]
```diff
@@ -360,12 +363,46 @@ def _get_on_demand_feature_view_workunit(
 
         return MetadataWorkUnit(id=on_demand_feature_view_name, mce=mce)
 
+    def _get_tags_and_owners(self, obj: Union[Entity, FeatureView, FeastField]) -> list:
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:366]: We written and defensive. Nice work!

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 6]
```diff
@@ -81,6 +88,14 @@
     feast.types.Invalid: MLFeatureDataType.UNKNOWN,
 }
 
+# FIXME: Update to have more owners
+_owner_mapping: Dict[str, Dict[str, Any]] = {
+    "Datahub": {
+        "owner_type": builder.OwnerType.USER,
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:94]: Does feast support GROUP owners as well? Thanks for calling this out!

Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py:94]: yes, we can pass the owner as a "Person" (user) or "Group"
Screenshot 2024-11-08 at 17 00 44

Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py:94]: the issue is that from Feast side I can only specify the owner: str (so, for example "Margarida") and this should map to a OwnerType.USER and to the appropriate OwnershipTypeClass.

Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py]: ok got it !

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 7]
```diff
@@ -216,9 +223,26 @@ def _get_entity_workunit(
 
         feature_view_name = f"{self.feature_store.project}.{feature_view.name}"
 
+        aspects = [StatusClass(removed=False)]
+
+        if entity.tags.get("name"):
+            tag: str = entity.tags.get("name")
```
[the remark this answers is not in the document]
Comment (jjoyce0510) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py]: I see - in that case this mapping needs to be externalized via configuration... We need a config where users can specify the Feast Owner Type > DataHub Owner Type mapping. Ideally, we'd have some default configuration values. Are there any default owner types in Feast?

[the remark this answers is not in the document]
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py]: Unfortunately, there is no default owner type, we can only specify the owner name

File: metadata-ingestion/tests/integration/feast/feast_repository_mces_golden.json [hunk 8]
```diff
@@ -11,6 +11,7 @@
                 },
                 {
                     "com.linkedin.pegasus2avro.ml.metadata.MLPrimaryKeyProperties": {
+                        "customProperties": {},
```
Comment (jjoyce0510) [inline metadata-ingestion/tests/integration/feast/feast_repository_mces_golden.json:14]: Does this test addition of owners + tags?

I don't see any tests for the new functionality..

Comment (jjoyce0510) [reply metadata-ingestion/tests/integration/feast/feast_repository_mces_golden.json:14]: I see below you have DataHub and some tags as well..

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 9]
```diff
@@ -360,6 +371,38 @@ def _get_on_demand_feature_view_workunit(
...
+            owners_aspect = OwnershipClass(owners=[owner_association])
+            aspects.append(owners_aspect)
+
+        return aspects
+
+    def _create_owner_association(self, owner: str) -> OwnerClass:
+
+        for mapping in self.source_config.owner_mappings:
+            if mapping["feast_owner_name"] == owner:
+                ownership_type_class: OwnershipTypeClass = mapping["ownership_type"]
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:400]: Minor: let's keep the naming consistent here:

datahub_owner_urn
datahub_ownership_type

Also, please add comments to the above Dict field to indicate which keys are required, and which are optional

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 10]
```diff
@@ -360,6 +371,38 @@ def _get_on_demand_feature_view_workunit(
...
+                return OwnerClass(
+                    owner=mapping["datahub_owner_urn"],
+                    type=ownership_type_class,
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:403]: What if a user does not provide an ownership type? 

Maybe we have some default here (TECHNICAL owner or something)

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 11]
```diff
@@ -91,6 +96,9 @@ class FeastRepositorySourceConfig(ConfigModel):
     environment: str = Field(
         default=DEFAULT_ENV, description="Environment to use when constructing URNs"
     )
+    owner_mappings: List[Dict[str, str]]  = Field(
+        default={}, description="Mapping of owner names to owner types"
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:100]: As mentioned below, we should have a better description of this argument with mentions of which keys are required, what happens if keys are not specified, etc.

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 12]
```diff
@@ -91,6 +96,9 @@ class FeastRepositorySourceConfig(ConfigModel):
     environment: str = Field(
         default=DEFAULT_ENV, description="Environment to use when constructing URNs"
     )
+    owner_mappings: List[Dict[str, str]]  = Field(
```
Comment (jjoyce0510) [inline metadata-ingestion/src/datahub/ingestion/source/feast.py:99]: Let's also add 2 additional flags:

- `enable_owner_extraction`: BOOL - If this is disabled, then we NEVER try to map owners. If this is enabled, then owner_mappings is REQUIRED to extract ownership. (We can use something like this for the docs)
- `enable_tag_extraction`: BOOL - If this is disabled, then we NEVER try to extract tags.

**Review by jjoyce0510 — COMMENTED** [review]
Left a few final comments. I think we are pretty close

File: metadata-ingestion/tests/integration/feast/feast_repository_mces_golden.json (hunk 8 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/tests/integration/feast/feast_repository_mces_golden.json:38]: updated now, thanks

File: metadata-ingestion/src/datahub/ingestion/source/feast.py (hunk 10 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py:403]: done, thanks

File: metadata-ingestion/src/datahub/ingestion/source/feast.py (hunk 9 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py:400]: I added type hints to reflect that they are optional and in case they are not passed they will have default values.

File: metadata-ingestion/src/datahub/ingestion/source/feast.py [hunk 13]
```diff
@@ -360,6 +371,38 @@ def _get_on_demand_feature_view_workunit(
...
+            owner_association = self._create_owner_association(owner)
+            owners_aspect = OwnershipClass(owners=[owner_association])
+            aspects.append(owners_aspect)
+
+        return aspects
+
+    def _create_owner_association(self, owner: str) -> OwnerClass:
+
+        for mapping in self.source_config.owner_mappings:
+            if mapping["feast_owner_name"] == owner:
```
[the remark this answers is not in the document]
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py:399]: done, thanks

File: metadata-ingestion/src/datahub/ingestion/source/feast.py (hunk 11 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py:100]: I added the comment and also added owner_mappings as Optional

File: metadata-ingestion/src/datahub/ingestion/source/feast.py (hunk 12 above)
Comment (margaridafernandes-trip) [reply metadata-ingestion/src/datahub/ingestion/source/feast.py]: done

**Review by jjoyce0510 — APPROVED** [review]
lgtm! thank you!
